rollback height map

arvidn opened the pull request:
When performing a reorg, we roll back existing blocks to an earlier state. This change makes us *also* roll back the height-to-hash map.

Before the height-to-hash refactor, we did also not roll this back, and I preserved this behaviour. However, there's some really subtle (and way too complex) logic in `get_block_generator()` ([here](https://github.com/Chia-Network/chia-blockchain/blob/main/chia/consensus/blockchain.py#L889)) that determines whether or not we are currently in a reorg based on whether a height-to-hash entry exists. This can (in rare instances) lead to thinking we are *not*, when we in fact are.

This addresses the "in-main-chain" issue gene experienced last week.

arvidn committed:
when rolling back the chain, also rollback the height-to-hash map

arvidn committed:
add test case for reorg with a block reference into the reorg-chain

arvidn commented on chia/consensus/blockchain.py:
```diff
@@ -380,6 +380,7 @@ async def _reconsider_peak(
         # Rollback sub_epoch_summaries
         self.__height_map.rollback(fork_height)
         await self.block_store.rollback(fork_height)
+        self._peak_height = None if fork_height < 0 else uint32(fork_height)
```
one possible issue with this is that we don't restore anything in case of a failure. but we don't restore the `__height_map` either.

mariano54 replied:
yeah that is a potential issue. In the case of reverting the TX, everything that happened needs to be rolled back

mariano54 replied:
Wouldn't this put the blockchain into an inconsistent state? the DB is not committed yet, so we still have all the coins and blocks there (not rolled back yet)

mariano54 replied:
Also shouldn't line 389 be:
`while fork_height >= 0` ?

arvidn replied:
> Wouldn't this put the blockchain into an inconsistent state? the DB is not committed yet, so we still have all the coins and blocks there (not rolled back yet)

Yes, it would. In fact, it *is* because the call to `rollback()` is never undone right now.

mariano54 reviewed (changes requested):
The desired behavior for receive_block is either it adds the block completely and finishes the reorg or does not add anything and keeps the peak identical. A key requirement of the blockchain is that it can only move forward in weight (you cannot end up with less weight than before). It looks like this is not the case, so it's a bug, but we should not move forward with trying to fix and settle the node into an intermediate state IMO.

arvidn committed:
if a reorg fails, restore height-to-hash and height-to-ses maps to their original state

arvidn committed:
restore _peak_height on reorg failure

arvidn commented:
@mariano54 I added two more commits that implements the restore behaviour, for when a reorg fails.

The most intrusive aspect is that I introduced an `ExitStack` context manager in the `_reconsider_peak()` function, meaning everything in there was indented one more level (please review with whitespace disabled).

The `ExitStack` is used to add the restore callback functions to, and then pop them at the end if we're successful.
